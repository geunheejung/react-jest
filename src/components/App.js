import React, { Component } from 'react';
import { Button } from 'reactstrap';
import Gift from './Gift';
import { max_number } from '../helper';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      gifts: []
    }
  }

  addGift = () => {
    const { gifts } = this.state;
    const ids = gifts.map(gift => gift.id);

    this.setState({
      gifts: [
        ...gifts,
        { id: max_number(ids) + 1 }
      ]
    });
  }

  removeGift = (id) => {
    const gifts = this.state.gifts.filter(gift => gift.id !== id);
    this.setState({
      gifts: gifts
    })
  }

  render() {
    return (
      <div>
        <h2>Gift Giver</h2>
        <Button
          className="btn-add"
          onClick={this.addGift}
        >Add Gift</Button>
        <div className="gift-list">
          {
            this.state.gifts.map(gift => (
              <Gift
                key={gift.id}
                gift={gift}
                removeGift={this.removeGift}
              />
            ))
          }
        </div>
      </div>
    )
  }
}

export default App;