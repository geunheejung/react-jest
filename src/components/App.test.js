import React, { Component } from 'react';
import { shallow } from 'enzyme';
import App from './App';
import requestAnimationFrame from '../tempPolyfills';
 
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter(), disableLifecycleMethods: true });

describe('App 컴포넌트 테스트', () => {
  const app = shallow(<App />);

  it('renders correctly', () => {
    expect(app).toMatchSnapshot();
  });

  it(`initalizes the 'state' with an empty list of gifts`, () => {
    expect(app.state().gifts).toEqual([]);
  });

  describe('when clicking the `add-gift` button', () => {
    const id = 1;

    beforeEach(() => {
      // 매 테스트 전에 실행됨
      app.find('.btn-add').simulate('click');
    });

    afterEach(() => {
      // 매 테스트 후에 실행됨
      app.setState({ gifts: [] });
    });

    it('adds a new gift to `state` when clicking the `add gift` button', () => {
      expect(app.state().gifts).toEqual([{ id }]);
    });

    it('adds a new gift to the rendered list when clicking the `add gift`', () => {
      expect(app.find('.gift-list').children().length).toEqual(1);
    });

    it('creates a Gift component', () => {
      expect(app.find('Gift').exists()).toBe(true);
    });

    describe(`and the user wants to remove the added gift`, () => {
      beforeEach(() => {
        app.instance().removeGift(id);
      });

      it('removes the gift from `state`', () => {
        expect(app.state().gifts).toEqual([]);
      });
    })
  });
});