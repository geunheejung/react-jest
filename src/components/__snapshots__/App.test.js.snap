// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 컴포넌트 테스트 renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Gift Giver
</h2>,
        <Button
          className="btn-add"
          color="secondary"
          onClick={[Function]}
          tag="button"
>
          Add Gift
</Button>,
        <div
          className="gift-list"
>
          
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Gift Giver",
        },
        "ref": null,
        "rendered": "Gift Giver",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Add Gift",
          "className": "btn-add",
          "color": "secondary",
          "onClick": [Function],
          "tag": "button",
        },
        "ref": null,
        "rendered": "Add Gift",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "gift-list",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Gift Giver
</h2>,
          <Button
            className="btn-add"
            color="secondary"
            onClick={[Function]}
            tag="button"
>
            Add Gift
</Button>,
          <div
            className="gift-list"
>
            
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Gift Giver",
          },
          "ref": null,
          "rendered": "Gift Giver",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Add Gift",
            "className": "btn-add",
            "color": "secondary",
            "onClick": [Function],
            "tag": "button",
          },
          "ref": null,
          "rendered": "Add Gift",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "gift-list",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
